source/reduce/operand_to_const_reduction_pass.cpp

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "operand_to_const_reduction_pass.h"
+#include "change_operand_reduction_opportunity.h"
+#include "source/opt/instruction.h"
+
+namespace spvtools {
+namespace reduce {
+
+using namespace opt;
+
+std::vector<std::unique_ptr<ReductionOpportunity>>
+OperandToConstReductionPass::GetAvailableOpportunities(
+    opt::IRContext* context) const {
+  std::vector<std::unique_ptr<ReductionOpportunity>> result;
+  assert(result.empty());
+  for (const auto& constant : context->GetConstants()) {
+    for (auto& function : *context->module()) {
+      for (auto& block : function) {
+        for (auto& inst : block) {
+          for (uint32_t i = 0; i < inst.NumOperands(); i++) {
```

Comment:
You can replace this loo and the check against operand type with:
```
inst.ForEachInId([&result,context](uint32_t* id_ptr) { /* code in the operand type if */});
```

Reply:
The problem I see here is that I use the loop counter "i" to know the index of the operand being changed when I create a ChangeOperandReductionOpportunity.

With the lambda approach I would get a pointer to the operand id, but wouldn't know (or at least not for free) which index this is at.

The choice seems to be to stick with the loop approach, or to identify an operand id in ChangeOperandReductionOpportunity by its pointer value.  I'm fine with either but slightly prefer the current index-based loop approach.  Let me know what you think.

Reply:
Resolving for now having added a comment to the code to explain, but do please unresolve if there's a better approach; I'm keen on a functional style where possible.

Reply:
In that case I would suggest checking code similar to the implementation of Instruction::WhileEachInId. Your current check misses SPV_OPERAND_TYPE_SCOPE_ID and SPV_OPERAND_TYPE_MEMORY_SEMANTICS_ID. These are required to be constants in the presence of the Shader capability, but they are effectively instruction ids.